lib/core/src/Cardano/Wallet/DB/Sqlite/AddressBook.hs

```diff
@@ -70,23 +72,36 @@ class AddressBookIso s where
     -- and its two components.
     addressIso :: Iso' s (Prologue s, Discoveries s)
 
+getPrologue :: AddressBookIso s => s -> Prologue s
+getPrologue = withIso addressIso $ \from _ -> fst . from
+
+getDiscoveries :: AddressBookIso s => s -> Discoveries s
+getDiscoveries = withIso addressIso $ \from _ -> snd . from
+
 {-------------------------------------------------------------------------------
     Sequential address book
 -------------------------------------------------------------------------------}
 -- piggy-back on SeqState existing instance, to simulate the same behavior.
-instance AddressBookIso (Seq.SeqState n k)
-    => AddressBookIso (Seq.SeqAnyState n k p)
+instance ( (key == SharedKey) ~ 'False, Eq (Seq.SeqState n key))
+    => AddressBookIso (Seq.SeqAnyState n key p)
   where
-    data Prologue (Seq.SeqAnyState n k p) = PS (Prologue (Seq.SeqState n k))
-    data Discoveries (Seq.SeqAnyState n k p) = DS (Discoveries (Seq.SeqState n k))
+    data Prologue (Seq.SeqAnyState n key p) = PS (Prologue (Seq.SeqState n key))
+    data Discoveries (Seq.SeqAnyState n key p) = DS (Discoveries (Seq.SeqState n key))
 
     addressIso = withIso addressIso $ \from to ->
         let from2 st = let (a,b) = from $ Seq.innerState st in (PS a, DS b)
             to2 (PS a, DS b) = Seq.SeqAnyState $ to (a,b)
         in  iso from2 to2
 
+instance Eq (Seq.SeqState n k) => Eq (Prologue (Seq.SeqAnyState n k p)) where
```

Comment:
I wonder if there is no way to kinda automate those equality instances in such a way that it could be derived? There is quite a number of them here